### PR TITLE
NOISSUE: adding remote terminal

### PR DESCRIPTION
**Pull request by mteodor**
Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -1545,6 +1545,25 @@
       "resolved": "https://registry.npmjs.org/amdefine/-/amdefine-1.0.1.tgz",
       "integrity": "sha1-SlKCrBZHKek2Gbz9OtFR+BfOkfU="
     },
+<<<<<<< HEAD
+=======
```
Comment (manuio) [inline package-lock.json]: Remove the merge tags and probably also the packages, at least jsoneditor

File: src/app/common/interfaces/models.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
+export interface User {
```
Comment (manuio) [inline src/app/common/interfaces/models.ts]: why this file? Use this one: https://github.com/mainflux/ui/blob/master/src/app/common/interfaces/mainflux.interface.ts

File: src/app/pages/things/gateways/details/xterm/gateways.xterm.component.ts [hunk 3]
```diff
@@ -1,65 +1,107 @@
...
+    if (state === MqttConnectionState.CONNECTED) {
+      this.notificationsService.success('Connected to MQTT broker', '');
+      this.subToChan();
+    }
+  }
+
+  subToChan() {
+    // var term: Terminal;
+    // term = new Terminal();
+    // term.open(this.terminalElement.nativeElement);
```
Comment (manuio) [inline src/app/pages/things/gateways/details/xterm/gateways.xterm.component.ts]: Should this be commented?

File: src/app/pages/things/gateways/details/xterm/gateways.xterm.component.ts [hunk 4]
```diff
@@ -1,65 +1,107 @@
...
   subscriptions: Subscription[] = new Array();
+  uuid: String;
 
-  public terminal: Terminal;
-  @ViewChild('terminal', { static: false }) terminalElement: ElementRef;
+  chanSub: Subscription;
+  stateSub: Subscription;
+  //public terminal: Terminal;
+
+ // @ViewChild('terminal', { static: false }) terminalElement: ElementRef;
```
Comment (manuio) [inline src/app/pages/things/gateways/details/xterm/gateways.xterm.component.ts]: rm comments

File: src/app/common/interfaces/mainflux.interface.ts [hunk 5]
```diff
@@ -11,6 +11,8 @@ export interface Channel {
   metadata?: any;
 }
 
+
+
```
Comment (manuio) [inline src/app/common/interfaces/mainflux.interface.ts]: rm this please

File: src/app/pages/things/gateways/details/xterm/gateways.xterm.component.html [hunk 6]
```diff
@@ -1 +1 @@
-<div #terminal id="terminal"></div>
+<div #terminal id="terminal"></div>
```
Comment (manuio) [inline src/app/pages/things/gateways/details/xterm/gateways.xterm.component.html]: No new line at the end of the file

File: src/app/pages/things/gateways/details/xterm/gateways.xterm.component.ts [hunk 7]
```diff
@@ -1,65 +1,111 @@
...
+
+  publish(channel: string, bn: string, n: string, vs: string) {
+    const topic = `${this.createTopic(channel)}/req`;
+    const t = Date.now() / 1000;
+    const payload = this.createPayload(bn, n, t, vs);
+    this.mqttService.publish(topic, payload).subscribe(e => {
+    },
+      err => {
+      this.notificationsService.error('Failed to publish', '${err}');
+    });
```
Comment (manuio) [inline src/app/pages/things/gateways/details/xterm/gateways.xterm.component.ts]: Please do:
```
    this.mqttService.publish(topic, payload).subscribe(
        resp => {
        },
        err => {
            this.notificationsService.error('Failed to publish', '${err}');
        },
    );
```

File: src/app/pages/things/gateways/details/xterm/gateways.xterm.component.ts [hunk 8]
```diff
@@ -1,65 +1,111 @@
...
+      this.mqttService.publish(topic, 'payload');
+      this.chanSub = this.mqttService.observe(topic).subscribe(
+        (message: IMqttMessage) => {
+          let res: string;
+          const pl = message.payload.toString();
+          res = JSON.parse(pl);
+          const msg = <Message>(<any>res[0]);
+          term.write(msg.vs);
+        });
+      this.notificationsService.success(`Subscribed to channel ${topic}`, '');
```
Comment (manuio) [inline src/app/pages/things/gateways/details/xterm/gateways.xterm.component.ts]: This should be called after the subscribe response, right?

File: src/app/pages/things/gateways/details/xterm/gateways.xterm.component.ts [hunk 9]
```diff
@@ -1,65 +1,111 @@
...
+  ) {
+    this.connected = false;
+    this.uuid = uuid();
+  }
+
+  ngOnChanges() {
+    if (this.gateway === undefined ||
+        this.terminal === undefined ||
+        this.connected === true)
+    return;
```
Comment (manuio) [inline src/app/pages/things/gateways/details/xterm/gateways.xterm.component.ts]: Wow, this works?

File: src/app/pages/things/gateways/details/xterm/gateways.xterm.component.ts [hunk 10]
```diff
@@ -1,65 +1,111 @@
...
     this.terminal = new Terminal();
     this.terminal.open(this.terminalElement.nativeElement);
     this.terminal.writeln('Welcome to Mainflux IoT Agent');
-    this.terminal.write('$ ');
-    this.terminal.on('data', function(data: string) {
-      this.mqttManagerService.publish(this.gateway.metadata.ctrlChannelID, '1', 'terminal', btoa(data));
-    }.bind(this));
+    this.terminal.onData( data => {
+      const vs = `c,${data}`;
+      this.publish(this.gateway.metadata.ctrlChannelID, this.uuid, 'term', btoa(vs));
```
Comment (manuio) [inline src/app/pages/things/gateways/details/xterm/gateways.xterm.component.ts]: Create a const for `term` please

File: src/app/pages/things/gateways/details/xterm/gateways.xterm.component.ts (hunk 9 above)
Comment (mteodor) [reply src/app/pages/things/gateways/details/xterm/gateways.xterm.component.ts:43]: why not?

**Pull request merged**